File: app/src/main/java/org/worshipsongs/component/HomeViewerPageAdapter.java
package org.worshipsongs.component;

import android.app.Activity;
import android.content.pm.ActivityInfo;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.util.Log;

import org.worshipsongs.WorshipSongApplication;
import org.worshipsongs.fragment.SongsFragment;
import org.worshipsongs.listener.SongContentViewListener;
import org.worshipsongs.registry.FragmentRegistry;
import org.worshipsongs.registry.ITabFragment;

import java.util.List;

/**
 * Author: madasamy.
 * version: 1.0.0
 */
public class HomeViewerPageAdapter extends FragmentPagerAdapter
{
    private static final String DEF_TYPE = "string";
    private Activity activity;
    private List<String> titles;
    private SongContentViewListener songContentViewListener;
    private FragmentRegistry fragmentRegistry = new FragmentRegistry();

    public HomeViewerPageAdapter(FragmentManager fragmentManager, Activity activity,
                                 List<String> titles, SongContentViewListener songContentViewListener)
    {
        super(fragmentManager);
        this.titles = titles;
        this.activity = activity;
        this.songContentViewListener = songContentViewListener;
    }

    @Override
    public Fragment getItem(int position)
    {
        ITabFragment fragment = fragmentRegistry.findByTitle(activity, titles.get(position));
        if (fragment != null) {
            fragment.setListenerAndBundle(songContentViewListener, null);
            return (Fragment) fragment;
        } else {
            return SongsFragment.newInstance(null);
        }
    }

    @Override
    public CharSequence getPageTitle(int position)
    {
        return "";
    }

    @Override
    public int getCount()
    {
        return titles.size();
    }

}

File: app/src/main/java/org/worshipsongs/component/HomeViewerPageAdapter.kt
package org.worshipsongs.component

import android.app.Activity
import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter

import org.worshipsongs.fragment.SongsFragment
import org.worshipsongs.listener.SongContentViewListener
import org.worshipsongs.registry.FragmentRegistry

/**
 * Author: madasamy.
 * version: 1.0.0
 */
class HomeViewerPageAdapter(fragmentManager: FragmentManager, private val activity: Activity, private val titles: List<String>, private val songContentViewListener: SongContentViewListener?) : FragmentPagerAdapter(fragmentManager)
{
    private val fragmentRegistry = FragmentRegistry()

    override fun getItem(position: Int): Fragment
    {
        Log.d(TAG, "Finding the fragment for the position $position")
        val fragment = fragmentRegistry.findByTitle(activity, titles[position])
        if (fragment != null)
        {
            fragment.setListenerAndBundle(songContentViewListener, Bundle())
            return fragment as Fragment
        } else
        {
            Log.d(TAG, "No fragment found for the position $position. Returning the Songs fragment instead...")
            return SongsFragment.newInstance(Bundle())
        }
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return ""
    }

    override fun getCount(): Int {
        return titles.size
    }

    companion object {
        val TAG = HomeViewerPageAdapter::class.simpleName
    }
}

